### PR TITLE
Update BigInt methods to use Chapel types instead of C types.

### PR DESCRIPTION
**Pull request by thomasvandoren**
Update BigInt methods to use Chapel types instead of C types. Then, inside
the methods, use safe_cast() to cast to appropriate C type. Below is the list
of C types and the corresponding Chapel types that replaced them.

Previously, the BigInt interface was not portable across 32- and 64-bit platforms.
This is due to the size of C types changing on those platforms, e.g. c_ulong is
one size for linux32 and another for linux64. The new interface is consistent for
both 32- and 64-bit platforms, _and_ it safely downcasts when the target C
type is smaller than the Chapel type. For example, when casting from a
64-bit chapel uint to a 32-bit c_ulong, there is a runtime check that confirms
the value won't overflow.

**C type => Chapel type**

`c_ulong` => `uint`
`c_long` => `int`
`c_int` => `int`
`size_t` => `uint`

And since doubles are the same size across platforms, update instances
of c_double to be real. Use regular cast to convert real to c_double.

I verified this works by running the following on linux32, darwin, and linux64:

```bash
start_test \
  test/studies/shootout/pidigits/ \
  test/modules/standard/gmp/ \
  test/release/examples/benchmarks/shootout/pidigits.chpl
```

### CODE REVIEW

**Comment by thomasvandoren** [comment]
test/modules/standard/gmp/studies/gmp-chudnovsky.chpl still fails on linux32, but that is not related to BigInt, so I'm not going to worry about.

**Comment by thomasvandoren** [comment]
@mppf -- will you please review?

**Comment by thomasvandoren** [comment]
I don't think you need to go over the changes with a fine tooth comb. I'm more interested in your thoughts on the spirit of this change.

In the future, we should write some more tests, since the vast majority of BigInt is not exercised in our testing system. :-\

File: test/studies/shootout/pidigits/thomasvandoren/pidigits-ledrug-BigInt.chpl [hunk 1]
```diff
@@ -56,8 +56,8 @@ iter genDigits(numDigits) {
   }
 
   // Helper function to extract the nth digit.
-  proc extractDigit(nth: c_ulong) {
-    tmp1.mul_ui(numer, nth);
+  proc extractDigit(nth: uint) {
+    tmp1.mul_ui(numer, safe_cast(c_ulong, nth));
```
Comment (mppf) [inline test/studies/shootout/pidigits/thomasvandoren/pidigits-ledrug-BigInt.chpl]: Why do you need safe_cast here? I thought mul_ui now takes in uint?

**Comment by mppf** [comment]
Looks good to me, except for the one test comment. Also, make sure to add a comment at the top of BigInt somewhere that describes what this all means (runtime failures if you pass a number larger than C's corresponding integer type for the GMP functions). I think that we could get rid of those run-time failures for GMP by creating a GMP number by shifting in those cases, but that's future work. A comment describing that would be nice. After that, go ahead and merge.

**Pull request merged**